### ISSUE TITLE
Doy you have any plan for image upload?

### ISSUE THREAD

**Issue by hesampour** [kolkov/angular-editor#5]
Hi
thanks for your great work. really a native angular 2+ WYSIWYG editor was needed. but if you add the image automatic upload it would be great. sth like what textbox.io have.

thanks

**Comment by kolkov** [comment]
Hi! Update to v.0.8.0 please and try with `config.uploadUrl.`
Your backend must return `imageUrl`.

**Comment by hesampour** [comment]
Thanks a lot for adding this excellent feature. but the problem is that the image is uploaded and the path but  no image is shown in the editor and just i see an empty <img> tag? 

And also. Another good feature that textbox.io have is that when we copy and paste the image it automatically uploads that. I think if you could also add this feature this is the best editor for angular.

**Comment by kolkov** [comment]
This workes for me:
`{
"status":true,
"originalName":"28279162_1980455641981763_1599997490751603080_n.jpg",
"generatedName":"88f1aa6a-a9ce-4a3c-ada5-51cdf4bdaadd",
"msg":"Image upload successful",
"imageUrl":"data/img/product/88f1aa6a-a9ce-4a3c-ada5-51cdf4bdaadd.jpg"
}`

**Comment by kolkov** [comment]
> And also. Another good feature that textbox.io have is that when we copy and paste the image it automatically uploads that. I think if you could also add this feature this is the best editor for angular.

There are no fundamental problems with this, the only problem is how beautiful it is to do it in the interface? I would also like to add the function of cropping and resizing images before unloading.
Please make a separate request for this function, that it would not get lost with us.
Thanks!

**Comment by hesampour** [comment]
{ "status":true, "originalName":"28279162_1980455641981763_1599997490751603080_n.jpg", "generatedName":"88f1aa6a-a9ce-4a3c-ada5-51cdf4bdaadd", "msg":"Image upload successful", "imageUrl":"data/img/product/88f1aa6a-a9ce-4a3c-ada5-51cdf4bdaadd.jpg" }

From here I found how should I return the success reply. thanks alot. 
instead of setting the upload url in config if there were a onUpload function that we could call after making a call request it would be great. 

Again thanks a lot for your great project. 
I think many developers needs such a native editor.

**Comment by kolkov** [comment]
I am very glad! Thanks!

**Comment by hesampour** [comment]
yes. we call a method in the api and send the image data to the function. the function in the api have access to save the data in the server and save the image and return the image url. with the returned url you can show the image.

**Comment by DruRio** [comment]
Can you bring me an example?

**Comment by hesampour** [comment]
[Authorize(Roles = "admin")]

        [ActionName("UploadBase64")]
        public OperationResult UploadBase64(base64 im) {
            try
            {
                var dirPath = Path.Combine(
                    HttpContext.Current.Server.MapPath("~/uploads/images/"),
                    im.imageFolder);
                if (!Directory.Exists(dirPath)) {
                    Directory.CreateDirectory(dirPath);
                }

                var path = Path.Combine(dirPath, im.fileName);
                int count = 0;
                Random rand = new Random(10000);
                while (File.Exists(path) && count<1000) {
                    count++;
                    var names = im.fileName.Split('.');
                    if (names.Count() > 1)
                    {
                        im.fileName = names[0] + rand.Next() + '.'+ names[1];
                    }
                    else {
                        im.fileName = names[0] + rand.Next();
                    }
                    path = Path.Combine(dirPath, im.fileName);
                }
                File.WriteAllBytes(path, Convert.FromBase64String(im.Image));
                OperationResult result = new OperationResult();
                result.succeeded = true;
                result.message = "uploads/images/" + im.imageFolder + '/' + im.fileName;
                return result;
            }
            catch(Exception ex) {
                OperationResult result = new OperationResult();
                result.succeeded = false;
                result.message = ex.ToString();
                return result;
            }
        }

and this is the the front service

    const binaryString = readerEvt.target.result;
    this.imageService.Uploadbase64(btoa(binaryString), this.fileName, this.imageFolder, this.url).subscribe(
      result => {
        if (result.succeeded) {
          this.imagePath = baseServer + '/' + result.message;
          this.imagePathChange.emit(this.imagePath);
        } else {
          this.toaster.error('خطایی در ذخیره سازی تصویر به وجود  آمده است');
        }
      },
      err => this.toaster.error(' خطای اتصال به سرور'));

**Comment by Sharanjaa** [comment]
How to add the function in Angular editor config in order to check file type whether its image?

**Comment by alusev** [comment]
How do I show an error in case if an image was not uploaded?

**Comment by maballo** [comment]
Can we change the method to PUT?

**Comment by kolkov** [comment]
> Can we change the method to PUT?
Hi!
Why we need to do this?

**Comment by nunos7** [comment]
> Hi,
> Does anyone have a complete example?
> I am not understanding how to implement the call to the service that uploads the image...

If we click on the upload image, before giving the focus to the editor, the image is uploaded but the image is not loaded in the editor. We have to first focus on the editor and then click on the button to insert the image.

**Comment by prakashksp20** [comment]
How to get image from angular 6+ WYSIWYG editor image upload only frontend?

**Comment by mdelarosa90** [comment]
Hi, How can I use the uploadUrl, if I'm using the service of Storage in Firebase ?

**Comment by Artanty** [comment]
Hello! Good editor! 
I can't understand how to customize image upload.

i get:
Request header field enctype is not allowed by Access-Control-Allow-Headers in preflight response.

**Comment by hesampour** [comment]
That's because of your server side. you should allow Access-Control-Allow in your server side implementation. What language did you used in your server side?

**Comment by Artanty** [comment]
php

<?php
header("Access-Control-Allow-Origin: *");
header("Access-Control-Allow-Methods: PUT, GET, POST, DELETE");
header("Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept");

**Comment by Artanty** [comment]
`<?php
header("Access-Control-Allow-Origin: *");
header("Access-Control-Allow-Methods: PUT, GET, POST, DELETE");
header("Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept");`

**Comment by qqaimh** [comment]
Hello. How to Add a url before the image src , just like below:
`<img src="http://47.103.42.176:17241/newsimage/newsimage/201602172159547321.jpg">`

In the code of my node.js server, I deal with the upload image ,just like below:
`let baseurl = 'http://47.103.42.176:17241/'
  ctx.set("Content-Type", "application/json");
  ctx.body = { 
            "status":true, 
            "originalName": file.name, 
            "generatedName": file.name, 
            "msg":"Image upload successful", 
            "imageUrl":baseurl+"newsimage/"+`${file.name}`, 
        }`

But the editor always insert image ,just like below:
`<img src="/newsimage/201602172159547321.jpg">`

 I use your editor htmlcontent in ionic4 /Angular to make my mobile application. 
if the  image can not have the base url, the show image will be......

**Comment by Ches1er** [comment]
to Artanty.
I hope you`ve found a solution!
If it is not too late:
In Angular
1. In config set uploadUrl which your api server will get, for example: 'http:\\mydomain\api\ang_editor\upload_image'
In PHP (I use Laravel)
2. Make api route:
Route::post('ang_editor\upload_image','Main\ApiImagesController@actionUploadImage');
3. In ApiImagesController make method which handle request:
public function actionUploadImage(Request $request){
        $file = $request->file('file');
        $uploaddir = 'img/';
        $file_name = $file -> getClientOriginalName();
        $destination_path = $uploaddir;
        if ($file->move($destination_path,$file_name)) {
            return json_encode($this->images_service->addImage($destination_path, $file_name));
        }
        else return json_encode(['response' => 'error']);

    }
4. Then make service which handle uploading process:
public function addImage($destination_path, $file_name)
    {
        $path = 'http://mydomain/'.$destination_path.'/'.$file_name;
        if (Image::create(['name'=>$file_name, 'path'=>$path]))return [
            'status' => true,
            'originalName' => $file_name,
            'generatedName' => $file_name,
            'msg'=>"Image upload successful",
            'imageUrl' => $path
        ];
        return ['response'=>'error'];
    }
And thats all, for my project it works very well. 
Hope in helps you!

**Comment by shivanshu2394** [comment]
I got error in the frontend when I hit API .
ERROR TypeError: Cannot read property 'nativeElement' of undefined
    at AngularEditorToolbarComponent.fileReset (kolkov-angular-editor.js:906)
    at SafeSubscriber._next (kolkov-angular-editor.js:881)
    at SafeSubscriber.__tryOrUnsub (Subscriber.js:185)
    at SafeSubscriber.next (Subscriber.js:124)
    at Subscriber._next (Subscriber.js:72)
    at Subscriber.next (Subscriber.js:49)
    at MergeMapSubscriber.notifyNext (mergeMap.js:69)
    at InnerSubscriber._next (InnerSubscriber.js:11)
    at InnerSubscriber.next (Subscriber.js:49)
    at XMLHttpRequest.onLoad (http.js:1961)

**Comment by kolkov** [comment]
@shivanshu2394 Hi! What version on packages tou are using? Angular Editor and others?

**Comment by shivanshu2394** [comment]
Hi, @kolkov , These are the versions i am using...
  "@angular/cli": "~8.1.1",
"@kolkov/angular-editor": "^1.0.0-beta.4".

Actually the image is displayed while uploading, but i am getting the error as well.

**Comment by kolkov** [comment]
@shivanshu2394 This works fine
` {
    "status":true,
    "originalName":"photo_2019-09-18_14-50-27.jpg",
    "extension":".jpg",
    "generatedName":"51e8789c-ce47-4828-aeaf-d3a7711fbf6e",
    "msg":"Image upload successful",
    "imageUrl":"data/img/product/51e8789c-ce47-4828-aeaf-d3a7711fbf6e.jpg"
}`

**Comment by shivanshu2394** [comment]
Hi @kolkov . I am sending base 64 from back end. The image upload is working fine. But i am getting an error as well..
I am returning this from my backend: 
 var imageUrl = "data:image/jpeg;base64,"+base64ofImage;

ERROR TypeError: Cannot read property 'nativeElement' of undefined
at AngularEditorToolbarComponent.fileReset (kolkov-angular-editor.js:906)

Why is this error coming anyways. ?

**Comment by kolkov** [comment]
@shivanshu2394 how you select image?

**Comment by shivanshu2394** [comment]
@kolkov 
 From the Angular Editor i am selecting image to be uploaded. 
The image is sent to my .net core web API. 
It is saved in a folder and converted into Base64.
I am returning the Base64 value from there.
Now, the image is displayed in the Editor. BUT THE ERROR MENTIONED ABOVE IS ALSO THERE.

I don't know what i am doing wrong.!

**Comment by payz0** [comment]
After set config.uploadUrl:'folder'. on console result ERROR TypeError: Cannot read property 'toLowerCase' of undefined.
And i'm still confuse how use { "status":true, "originalName":"28279162_1980455641981763_1599997490751603080_n.jpg", "generatedName":"88f1aa6a-a9ce-4a3c-ada5-51cdf4bdaadd", "msg":"Image upload successful", "imageUrl":"data/img/product/88f1aa6a-a9ce-4a3c-ada5-51cdf4bdaadd.jpg" }
in angular 8

**Comment by kolkov** [comment]
@payz0 config.uploadUrl is an endpoint on your server side or in the oter internet resource. And then your server response with imageUrl after successful store uploaded image.

**Comment by MurrayFurtado** [comment]
I am seeing this  ERROR TypeError: Cannot read property 'toLowerCase' of undefined error too. Same distro of angular-editor, v8.22 of angular.

**Comment by rajesh1107** [comment]
@kolkov i didn't find complete documentation of this(specifically for editor config settings), can you please let me know the documentation link?

**Comment by Artanty** [comment]
> to Artanty.
> I hope you`ve found a solution!
> If it is not too late:
> In Angular
> 
> 1. In config set uploadUrl which your api server will get, for example: 'http:\mydomain\api\ang_editor\upload_image'
>    In PHP (I use Laravel)
> 2. Make api route:
>    Route::post('ang_editor\upload_image','Main\ApiImagesController@actionUploadImage');
> 3. In ApiImagesController make method which handle request:
>    public function actionUploadImage(Request $request){
>    $file = $request->file('file');
>    $uploaddir = 'img/';
>    $file_name = $file -> getClientOriginalName();
>    $destination_path = $uploaddir;
>    if ($file->move($destination_path,$file_name)) {
>    return json_encode($this->images_service->addImage($destination_path, $file_name));
>    }
>    else return json_encode(['response' => 'error']);
>    }
> 4. Then make service which handle uploading process:
>    public function addImage($destination_path, $file_name)
>    {
>    $path = 'http://mydomain/'.$destination_path.'/'.$file_name;
>    if (Image::create(['name'=>$file_name, 'path'=>$path]))return [
>    'status' => true,
>    'originalName' => $file_name,
>    'generatedName' => $file_name,
>    'msg'=>"Image upload successful",
>    'imageUrl' => $path
>    ];
>    return ['response'=>'error'];
>    }
>    And thats all, for my project it works very well.
>    Hope in helps you!


Ches1er , thank you!
I use Laravel too.
I made 1, 2, 3 steps and now images saving works,
but I stopped at step 4.
I made service:
+++++++
namespace App\Classes;
  
class ImagesService {
	
    public function addImage($destination_path, $file_name)
    {
    
    	$path = 'http://localhost:8000/'.$destination_path.'/'.$file_name;
	
    	if (Image::create(['name'=>$file_name, 'path'=>$path]))return [
    	'status' => true,
    	'originalName' => $file_name,
    	'generatedName' => $file_name,
    	'msg'=>"Image upload successful",
    	'imageUrl' => $path
    	];
    	return ['response'=>'error'];
    }
}
++++++

And I can't connect it to image controller(step 3).

response while upload image:
"message": "Undefined property: App\\Http\\Controllers\\imageUpload::$images_service"

Can you help me?)

**Comment by Ches1er** [comment]
Hello, you can skip service if you wish, and do all process in controller. Just add code from service to controller, and modify it a little bit. But it is not a good practice. It always better to use dependency injection (as I did in my case). If you havent faced with dependency injection yet, I can try to explain using this case.

**Comment by Ches1er** [comment]
> > to Artanty.
> > I hope you`ve found a solution!
> > If it is not too late:
> > In Angular
> > 
> > 1. In config set uploadUrl which your api server will get, for example: 'http:\mydomain\api\ang_editor\upload_image'
> >    In PHP (I use Laravel)
> > 2. Make api route:
> >    Route::post('ang_editor\upload_image','Main\ApiImagesController@actionUploadImage');
> > 3. In ApiImagesController make method which handle request:
> >    public function actionUploadImage(Request $request){
> >    $file = $request->file('file');
> >    $uploaddir = 'img/';
> >    $file_name = $file -> getClientOriginalName();
> >    $destination_path = $uploaddir;
> >    if ($file->move($destination_path,$file_name)) {
> >    return json_encode($this->images_service->addImage($destination_path, $file_name));
> >    }
> >    else return json_encode(['response' => 'error']);
> >    }
> > 4. Then make service which handle uploading process:
> >    public function addImage($destination_path, $file_name)
> >    {
> >    $path = 'http://mydomain/'.$destination_path.'/'.$file_name;
> >    if (Image::create(['name'=>$file_name, 'path'=>$path]))return [
> >    'status' => true,
> >    'originalName' => $file_name,
> >    'generatedName' => $file_name,
> >    'msg'=>"Image upload successful",
> >    'imageUrl' => $path
> >    ];
> >    return ['response'=>'error'];
> >    }
> >    And thats all, for my project it works very well.
> >    Hope in helps you!
> 
> Ches1er , thank you!
> I use Laravel too.
> I made 1, 2, 3 steps and now images saving works,
> but I stopped at step 4.
> I made service:
> +++++++
> namespace App\Classes;
> 
> class ImagesService {
> 
> ```
> public function addImage($destination_path, $file_name)
> {
> 
> 	$path = 'http://localhost:8000/'.$destination_path.'/'.$file_name;
> 
> 	if (Image::create(['name'=>$file_name, 'path'=>$path]))return [
> 	'status' => true,
> 	'originalName' => $file_name,
> 	'generatedName' => $file_name,
> 	'msg'=>"Image upload successful",
> 	'imageUrl' => $path
> 	];
> 	return ['response'=>'error'];
> }
> ```
> 
> }
> ++++++
> 
> And I can't connect it to image controller(step 3).
> 
> response while upload image:
> "message": "Undefined property: App\Http\Controllers\imageUpload::$images_service"
> 
> Can you help me?)

Hello, you can skip service if you wish, and do all process in controller. Just add code from service to controller, and modify it a little bit. But it is not a good practice. It always better to use dependency injection (as I did in my case). If you havent faced with dependency injection yet, I can try to explain using this case.

**Comment by Artanty** [comment]
Hello, Ches1er !
I'm new to this, you explain me pls using this case)

**Comment by Ches1er** [comment]
> Hello, Ches1er !
> I'm new to this, can you explain me pls using this case)

Create “Contracts” folder, where we’ll store our contracts.
In this folder create interface - “ServiceApiImages”

interface ServiceApiImages
{
public function addImage($destination_path, $file_name);
}

Create DBApiImagesService which implements our contract

namespace App\Services\Main;

use App\Contracts\ServiceApiImages;
use App\Models\Image;

class DBApiImagesService implements ServiceApiImages
{

    public function addImage($destination_path, $file_name)
   {
    $path = ‘http://mydomain/’.$destination_path.'/'.$file_name;
    if (Image::create(['name'=>$file_name, 'path'=>$path]))return [
        'status' => true,
        'originalName' => $file_name,
        'generatedName' => $file_name,
        'msg'=>"Image upload successful",
        'imageUrl' => $path
    ];
    return ['response'=>'error'];
}
}

Via console create a provider 
Php artisan make:provider ServiceApiImagesProvider
Inside this provider realize only one method
public function register()
    {
        $this->app->bind(ServiceApiImages::class, DBApiImagesService::class);
    }
In this method we are binding our Contract with Service,  then in controllers we can inject Contract.  This approach very convenient.  If in future we’ll change our  DB source (from MySQL to smth else) what we have to do - is just create a new Service, and inside method “register” replace DBApiImagesService::class by our new service, and it works.
Then in Providers/ApiServiceProvider.php (you have this file form the box) register our new Provider.
public function register()
    {
        
        $this->app->register(ServiceApiImagesProvider::class);
    }
And in file Config/app.php we must add our new provider to the ‘providers’ array
In my case it looks like this
 App\Providers\Custom\Main\ServiceApiImagesProvider::class
And finally we can use our Service through Dependency Injection in every controller we want, just add it by constructor.

class ApiImagesController extends Controller
{
    protected $images_service;

    /**
     * ApiImagesController constructor.
     * @param ServiceApiImages $imgService
     */
    public function __construct(ServiceApiImages $imgService)
    {
        $this->images_service = $imgService;
    }
    public function actionUploadImage(Request $request){
        $file = $request->file('file');
        $uploaddir = 'img';
        $file_name = $file -> getClientOriginalName();
        $destination_path = $uploaddir;
        if ($file->move($destination_path,$file_name)) {
            return json_encode($this->images_service->addImage($destination_path, $file_name));
        }
        else return json_encode(['response' => 'error']);
    }
}
I suggest you to read official laravell documentation about: 
https://laravel.com/docs/5.8/container
https://laravel.com/docs/5.8/providers
Dependency Injection - is a very powerful tool.

**Comment by anderking** [comment]
I am using this api: https://api.kraken.io/v1/upload that I found on a page to test the load using the property uploadUrl But this error returns me too


core.js:9110 ERROR TypeError: Cannot read property 'toLowerCase' of undefined
    at HttpXsrfInterceptor.intercept (http.js:2190)
    at HttpInterceptorHandler.handle (http.js:1471)
    at HttpInterceptingHandler.handle (http.js:2255)
    at MergeMapSubscriber.project (http.js:1246)
    at MergeMapSubscriber._tryNext (mergeMap.js:46)
    at MergeMapSubscriber._next (mergeMap.js:36)
    at MergeMapSubscriber.next (Subscriber.js:49)
    at Observable._subscribe (scalar.js:4)
    at Observable._trySubscribe (Observable.js:42)
    at Observable.subscribe (Observable.js:28)

"@kolkov/angular-editor": "^1.0.2",

**Comment by kolkov** [comment]
@anderking provide an repo please with this error.

**Comment by anderking** [comment]
@kolkov https://github.com/anderking/angular-editor-anderking

**Comment by NRDeshpande** [comment]
Hi, @kolkov can you please tell how you this solved problem, I am getting the same error now.

instrument.js:110 ERROR TypeError: Cannot read property 'toLowerCase' of undefined
    at HttpXsrfInterceptor.intercept (http.js:2790)
    at HttpInterceptorHandler.handle (http.js:1941)
    at HttpInterceptingHandler.handle (http.js:2873)
    at MergeMapSubscriber.project (http.js:1665)
    at MergeMapSubscriber._tryNext (mergeMap.js:46)
    at MergeMapSubscriber._next (mergeMap.js:36)
    at MergeMapSubscriber.next (Subscriber.js:49)
    at Observable._subscribe (subscribeToArray.js:3)
    at Observable._trySubscribe (Observable.js:42)
    at Observable.subscribe (Observable.js:28)
    at MergeMapOperator.call (mergeMap.js:21)
    at Observable.subscribe (Observable.js:23)
    at AngularEditorToolbarComponent.onFileChanged (kolkov-angular-editor.js:1160)
    at AngularEditorToolbarComponent_div_0_Template_input_change_61_listener (kolkov-angular-editor.js:157)

**Comment by kolkov** [comment]
@NRDeshpande hi! What versions of codebase you are using?

**Comment by vishalmote** [comment]
> @NRDeshpande hi! What versions of codebase you are using?

"@kolkov/angular-editor": "^1.1.2",

**Comment by vishalmote** [comment]
@kolkov , I am using angular 8, and "@kolkov/angular-editor": "^1.1.2",

following is my configuration
```

this.editorConfig = { // Check the interface for more options
            editable: true,
            spellcheck: true,
            height: 'auto',
            minHeight: '300px',
            maxHeight: 'auto',
            width: 'auto',
            minWidth: '0',
            translate: 'yes',
            enableToolbar: true,
            showToolbar: true,
            placeholder: '',
            defaultParagraphSeparator: 'p',
            defaultFontName: '',
            defaultFontSize: '',
            fonts: [
                { class: 'arial', name: 'Arial' },
                { class: 'times-new-roman', name: 'Times New Roman' },
                { class: 'calibri', name: 'Calibri' },
                { class: 'comic-sans-ms', name: 'Comic Sans MS' }
            ],
            customClasses: [
                {
                    name: 'quote',
                    class: 'quote',
                },
                {
                    name: 'redText',
                    class: 'redText'
                },
                {
                    name: 'titleText',
                    class: 'titleText',
                    tag: 'h1',
                },
            ],
            uploadUrl: 'http://localhost:9000/uploadMessageMedia',
            uploadWithCredentials: false,
            sanitize: true,
            toolbarPosition: 'top',
            toolbarHiddenButtons: [
                [],
                ['insertVideo']
            ]
        };
```
following is my html
```
<angular-editor [(ngModel)]="formData.body" [config]="editorConfig">
        </angular-editor>
```

my server side code
```


$router->post('/uploadMessageMedia', [
    'as' => 'uploadMessageMedia', 'uses' => 'MediaController@uploadMessageMedia'
]);

public function uploadMessageMedia(Request $request)
    {
        $file = $request->file('file');
        $uploaddir = 'img/';
        $file_name = $file->getClientOriginalName();
        $destination_path = $uploaddir;
        if ($file->move($destination_path, $file_name)) {
            return json_encode($this->addImage($destination_path, $file_name));
        } else return json_encode(['response' => 'error']);
    }

    public function addImage($destination_path, $file_name)
    {
        $path = 'http://localhost:9000/' . $destination_path . '/' . $file_name;
        return [
            'status' => true,
            'originalName' => $file_name,
            'generatedName' => $file_name,
            'msg' => "Image upload successful",
            'imageUrl' => $path
        ];
        // return ['response' => 'error'];
    }
```
I am able to successfully upload with help of above comments but lot of struggle need to be done by understanding each and every comment. I am giving above code for someone else help.

**But still I am facing two issue :-**

1. I have to focus the editor first then able to upload a file, If I didn't focus on editor and click image button and select a image, I am getting a following error

```
core.js:6014 ERROR TypeError: Cannot read property 'toLowerCase' of undefined
    at HttpXsrfInterceptor.intercept (http.js:2742)
    at HttpInterceptorHandler.handle (http.js:1921)
    at HttpInterceptingHandler.handle (http.js:2819)
    at MergeMapSubscriber.project (http.js:1648)
    at MergeMapSubscriber._tryNext (mergeMap.js:46)
    at MergeMapSubscriber._next (mergeMap.js:36)
    at MergeMapSubscriber.next (Subscriber.js:49)
    at Observable._subscribe (scalar.js:4)
    at Observable._trySubscribe (Observable.js:42)
    at Observable.subscribe (Observable.js:28)
```

2. once image get uploaded and inserted into editor,  following error is throwing

```
core.js:6014 ERROR TypeError: Cannot read property 'nativeElement' of undefined
    at AngularEditorToolbarComponent.fileReset (kolkov-angular-editor.js:909)
    at SafeSubscriber._next (kolkov-angular-editor.js:884)
    at SafeSubscriber.__tryOrUnsub (Subscriber.js:185)
    at SafeSubscriber.next (Subscriber.js:124)
    at Subscriber._next (Subscriber.js:72)
    at Subscriber.next (Subscriber.js:49)
    at MergeMapSubscriber.notifyNext (mergeMap.js:69)
    at InnerSubscriber._next (InnerSubscriber.js:11)
    at InnerSubscriber.next (Subscriber.js:49)
    at CatchSubscriber._next (Subscriber.js:72)
```


Please suggest.

**Comment by vishalmote** [comment]
> @kolkov , I am using angular 8, and "@kolkov/angular-editor": "^1.1.2",
> 
> following is my configuration
> 
> ```
> 
> this.editorConfig = { // Check the interface for more options
>             editable: true,
>             spellcheck: true,
>             height: 'auto',
>             minHeight: '300px',
>             maxHeight: 'auto',
>             width: 'auto',
>             minWidth: '0',
>             translate: 'yes',
>             enableToolbar: true,
>             showToolbar: true,
>             placeholder: '',
>             defaultParagraphSeparator: 'p',
>             defaultFontName: '',
>             defaultFontSize: '',
>             fonts: [
>                 { class: 'arial', name: 'Arial' },
>                 { class: 'times-new-roman', name: 'Times New Roman' },
>                 { class: 'calibri', name: 'Calibri' },
>                 { class: 'comic-sans-ms', name: 'Comic Sans MS' }
>             ],
>             customClasses: [
>                 {
>                     name: 'quote',
>                     class: 'quote',
>                 },
>                 {
>                     name: 'redText',
>                     class: 'redText'
>                 },
>                 {
>                     name: 'titleText',
>                     class: 'titleText',
>                     tag: 'h1',
>                 },
>             ],
>             uploadUrl: 'http://localhost:9000/uploadMessageMedia',
>             uploadWithCredentials: false,
>             sanitize: true,
>             toolbarPosition: 'top',
>             toolbarHiddenButtons: [
>                 [],
>                 ['insertVideo']
>             ]
>         };
> ```
> 
> following is my html
> 
> ```
> <angular-editor [(ngModel)]="formData.body" [config]="editorConfig">
>         </angular-editor>
> ```
> 
> my server side code
> 
> ```
> 
> 
> $router->post('/uploadMessageMedia', [
>     'as' => 'uploadMessageMedia', 'uses' => 'MediaController@uploadMessageMedia'
> ]);
> 
> public function uploadMessageMedia(Request $request)
>     {
>         $file = $request->file('file');
>         $uploaddir = 'img/';
>         $file_name = $file->getClientOriginalName();
>         $destination_path = $uploaddir;
>         if ($file->move($destination_path, $file_name)) {
>             return json_encode($this->addImage($destination_path, $file_name));
>         } else return json_encode(['response' => 'error']);
>     }
> 
>     public function addImage($destination_path, $file_name)
>     {
>         $path = 'http://localhost:9000/' . $destination_path . '/' . $file_name;
>         return [
>             'status' => true,
>             'originalName' => $file_name,
>             'generatedName' => $file_name,
>             'msg' => "Image upload successful",
>             'imageUrl' => $path
>         ];
>         // return ['response' => 'error'];
>     }
> ```
> 
> I am able to successfully upload with help of above comments but lot of struggle need to be done by understanding each and every comment. I am giving above code for someone else help.
> 
> **But still I am facing two issue :-**
> 
> 1. I have to focus the editor first then able to upload a file, If I didn't focus on editor and click image button and select a image, I am getting a following error
> 
> ```
> core.js:6014 ERROR TypeError: Cannot read property 'toLowerCase' of undefined
>     at HttpXsrfInterceptor.intercept (http.js:2742)
>     at HttpInterceptorHandler.handle (http.js:1921)
>     at HttpInterceptingHandler.handle (http.js:2819)
>     at MergeMapSubscriber.project (http.js:1648)
>     at MergeMapSubscriber._tryNext (mergeMap.js:46)
>     at MergeMapSubscriber._next (mergeMap.js:36)
>     at MergeMapSubscriber.next (Subscriber.js:49)
>     at Observable._subscribe (scalar.js:4)
>     at Observable._trySubscribe (Observable.js:42)
>     at Observable.subscribe (Observable.js:28)
> ```
> 
> 1. once image get uploaded and inserted into editor,  following error is throwing
> 
> ```
> core.js:6014 ERROR TypeError: Cannot read property 'nativeElement' of undefined
>     at AngularEditorToolbarComponent.fileReset (kolkov-angular-editor.js:909)
>     at SafeSubscriber._next (kolkov-angular-editor.js:884)
>     at SafeSubscriber.__tryOrUnsub (Subscriber.js:185)
>     at SafeSubscriber.next (Subscriber.js:124)
>     at Subscriber._next (Subscriber.js:72)
>     at Subscriber.next (Subscriber.js:49)
>     at MergeMapSubscriber.notifyNext (mergeMap.js:69)
>     at InnerSubscriber._next (InnerSubscriber.js:11)
>     at InnerSubscriber.next (Subscriber.js:49)
>     at CatchSubscriber._next (Subscriber.js:72)
> ```
> 
> Please suggest.


I have added some hack solution for problem 1 in above comment, but problem two still persist, check below code. this code will only work if there is only one angular-editor in page.
```
ngAfterViewInit() {
        document.getElementById("insertImage-").addEventListener("click", this.focusMe);
    }

    public focusMe() {
        document.getElementsByClassName("angular-editor-textarea")[0].focus();
    }
```

**Comment by vishalmote** [comment]
@kolkov ,


I have another requirement in this library, you library is great, Please check can following requirement be  putted :- 

**can you change api response acceptance format from :-** 

`
{ "status":true, "originalName":"photo_2019-09-18_14-50-27.jpg", "extension":".jpg", "generatedName":"51e8789c-ce47-4828-aeaf-d3a7711fbf6e", "msg":"Image upload successful", "imageUrl":"data/img/product/51e8789c-ce47-4828-aeaf-d3a7711fbf6e.jpg" }`


**To**


`
{ "status":true, "originalName":"photo_2019-09-18_14-50-27.jpg", "extension":".jpg", "generatedName":"51e8789c-ce47-4828-aeaf-d3a7711fbf6e", "msg":"Image upload successful", "htmlResponse":'<img src="data/img/product/51e8789c-ce47-4828-aeaf-d3a7711fbf6e.jpg"/>' }`


Because user may wanted to set some css to image or wrap anchor tag against the image.

**Comment by asifisthiaq24** [comment]
> Hi, @kolkov can you please tell how you solved this problem, I am getting the same error now.
> 
> instrument.js:110 ERROR TypeError: Cannot read property 'toLowerCase' of undefined
> at HttpXsrfInterceptor.intercept (http.js:2790)
> at HttpInterceptorHandler.handle (http.js:1941)
> at HttpInterceptingHandler.handle (http.js:2873)
> at MergeMapSubscriber.project (http.js:1665)
> at MergeMapSubscriber._tryNext (mergeMap.js:46)
> at MergeMapSubscriber._next (mergeMap.js:36)
> at MergeMapSubscriber.next (Subscriber.js:49)
> at Observable._subscribe (subscribeToArray.js:3)
> at Observable._trySubscribe (Observable.js:42)
> at Observable.subscribe (Observable.js:28)
> at MergeMapOperator.call (mergeMap.js:21)
> at Observable.subscribe (Observable.js:23)
> at AngularEditorToolbarComponent.onFileChanged (kolkov-angular-editor.js:1160)
> at AngularEditorToolbarComponent_div_0_Template_input_change_61_listener (kolkov-angular-editor.js:157)

To solve this issue use this in your html:
    <angular-editor [(ngModel)]="htmlContent" [config]="editorConfig" #myAngularEditor> 
    </angular-editor>{{ myAngularEditor.focus() }}
instead of this:
    <angular-editor [(ngModel)]="htmlContent" [config]="editorConfig" ></angular-editor>

**Comment by TrofinSorin** [comment]
@kolkov the nativeElement of undefined error has been fixed with release 1.1.3, but the Cannot read property 'toLowerCase' of undefined error remains, when you upload, it can't even reach to make the API Call to the uploadUrl

**Comment by Ruby-Thakur** [comment]
I tried to use server URL in uploadUrl but still its give me  ERROR TypeError: Cannot read property 'toLowerCase' of undefined
